src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+ * they are currently generated to comply with glibc rather than musl.
+ *
+ * Future plans:
+ *  - Compile cFE with musl and fix stat and dirent structures
+ *  - modularize into a bare bones filesystem library
+ *  - Refactor rest of code into posix compatible filesystem built on bare bones fs
+ *
  */
 
 /******************************************************************************
-** Tar Level Methods
+** fsobj Level Methods
 ******************************************************************************/
 
-/*
- * Loads the position in memory of linked tar file system
- * Checks for badly linked or no linked tar file.  The names given by
- * the linker are non-intuative so a description of error checking is given:
- * First checks to make sure that symbols have been overwritten by linking
- * process.  Next checks that the size is greater than 0.  Finally checks that
- * the end of the tar is after the start
- */
-uint32 tar_load()
+int32 chk_fd(int32 FD)
 {
-    if (!_binary_cFEfs_tar_start)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-    if (! &_binary_cFEfs_tar_size)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-    if (&_binary_cFEfs_tar_end < &_binary_cFEfs_tar_start)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-    size    = (size_t) &_binary_cFEfs_tar_size;
-    start   = &_binary_cFEfs_tar_start;
-    end     = &_binary_cFEfs_tar_end;
-    return  OS_FS_SUCCESS;
+    if (FD > OS_MAX_NUM_OPEN_FILES) return OS_FS_ERR_INVALID_FD;
+    if (FD <= 0)                        return OS_FS_ERR_INVALID_FD;
+
+    struct fd *filedes = &fd_tbl[FD];
+    if (!filedes->file) return OS_FS_ERR_INVALID_FD;
+    if (filedes->ino != filedes->file->ino) return OS_FS_ERR_INVALID_FD;
+    return OS_FS_SUCCESS;
 }
 
-/*
- * parses a loaded tar into fs[0]
- */
-uint32 tar_parse()
+//finds the next free file
+uint32 newfile_get(struct fsobj **o)
 {
-    assert(start && end);
-    assert(size < INT32_MAX);
-    assert(end - start > 0);
-    assert(size == (size_t) (end - start));
-    assert(fs[0].root->filedes == 1); //first fs should be rooted by first file
+    uint32 count = 0;
+    while (count < MAX_NUM_FILES && files[count].ino) {
+        count++;
+    }
+    if (count == MAX_NUM_FILES) return OS_FS_ERROR;
+    *o =  &files[count];
 
-    if (!fs[0].devname) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    **o = (struct fsobj) {
+        //ino needs to be unique and nonzero, so ino is defined as index+1
+        .ino = count +1
+    };
+    return OS_FS_SUCCESS;
+}
 
-    struct fsobj *o = fs[0].root;
-    uint32 offset = 0;
+uint32 file_insert(struct fsobj *o, char *path)
+{
+    //paths should always begin with '/' but we do not need it here
+    if (path[0] != '/') return OS_FS_ERR_PATH_INVALID;
+    path++;
 
-    while (offset + start < end) {
-        //tar ends after two empty records
-        if ( !(offset + start)[0] && !(offset + start)[TAR_BLOCKSIZE]) return OS_FS_SUCCESS;
+    assert(o && path && openfs);
```

Comment:
I am in a weird place.  That checking is done in the wrapper methods (as it should be), and redundant code is not good.  When this is a library it will return error codes, and the cFE will interpret them to return the cFE error codes.  char copy_buffer[F_PART_DATA_SIZE];